### ISSUE TITLE
Refactor cache system

### ISSUE THREAD

**Issue by neo22s** [koseven/koseven#283]
Cache system is not working as it's supposed. 

Read by @errotan 
https://github.com/koseven/koseven/issues/275#issuecomment-438763542

Key points:
- Cache system from the start of app to the end
- Using the cache system selected

Related:
https://github.com/koseven/koseven/issues/275
https://github.com/koseven/koseven/issues/192
https://github.com/koseven/koseven/issues/24

**Comment by toitzi** [comment]
anyone already into it?

**Comment by toitzi** [comment]
@neo22s How do you think we can solve this?
As @errotan already mentioned it is because Modules are not loaded when Kohana::init is called,
do you hink maybe integrating caching (since it is used by koseven core system) into the core would be the solution?

**Comment by neo22s** [comment]
@toitzi yes for me the cache system should be in the core and not as a module. Then when we init:

Koseven::init(array(
    'base_url'  => 'whatever.com',
    'errors'    => TRUE,
    'profile'   => FALSE,
    'caching'   => TRUE,
));

If they set it false its not used...

This is how I think it right now. But I am on a mobile phone and I am in the beach with caipirinhas so do not trust me. xD

**Comment by toitzi** [comment]
OK, finally i got a bit of time to take over on that issue. I am going to integrate the cache module into the core system, will take a lot of steps to fully update this in the code and a lot of unittests to be rewritten and added. Please be patient with me 😃

**Comment by roman-huliak** [comment]
is there any progress or this is dead issue?

**Comment by toitzi** [comment]
@roman-huliak  There is progress going on, i hope i can finish/merge it soon!

**Comment by lerman** [comment]
Also interested in the status of this. With caching enabled, my directory specific database.php (app/config/production/database.php) is not read correctly. I assume related?

**Comment by lerman** [comment]
Actually my above issue seems to be a different bug. We are caching the system configs and not allowing app ones to be loaded at all!

**Comment by Sebjugate** [comment]
If the solution to this is to integrate the cache into the core system, how would one utilize a custom cache driver?
We'd encounter the same problem, since unless that custom driver is also integrated into the core system, it would not be usable at `init()`.

**Comment by anna-springer** [comment]
@Sebjugate what we mean with integrate is more like shipping it together with koseven not as external module. Back when it was kohana all modules got shipped with it. Koseven later on removed them and made them ship seperate, so you would only need the code for modules you need.

The problem we discovered is that core classes on bootstrapping (before modules are included) use caching. No matter what you would specify as cache driver it would always use "File" as driver, which is why the "cache" folder was required to exist, even if some does not use "File" as driver.

So what we do is move that internal file caching to the "Cache" Class and ship the cache module as system module. Custom cache drivers will still work, the same way, this should almost be a non breaking change

**Comment by toitzi** [comment]
@Sebjugate what we mean with integrate is more like shipping it together with koseven not as external module. Back when it was kohana all modules got shipped with it. Koseven later on removed them and made them ship seperate, so you would only need the code for modules you need.

The problem we discovered is that core classes on bootstrapping (before modules are included) use caching. No matter what you would specify as cache driver it would always use "File" as driver, which is why the "cache" folder was required to exist, even if some does not use "File" as driver.

So what we do is move that internal file caching to the "Cache" Class and ship the cache module as system module. Custom cache drivers will still work, the same way, this should almost be a non breaking change

**Comment by Sebjugate** [comment]
@toitzi, Thanks for the response. First, let me preface my comment by saying it is very possible that I'm totally misunderstanding the problem and solution, so this is not to detract from the work or thought that has already been put into it by people who know much more and are much smarter than I am (hi @neo22s)!

But here is my understanding:

`cache` was always a module, even back to Kohana (See https://github.com/kohana/cache). However, at that time, it was *only* possible to use a baked-in file caching for the `find_files()` method via the `Kohana::cache` function, which did not utilize the `Cache` class at all. (See https://github.com/kohana/core/blob/bdbe81afb5a09cee4269d2e2210a0d293265231a/classes/Kohana/Core.php#L856)

My understanding of the underlying problem is that `Kohana::cache` was changed to use the the Cache module if it was available (issue #24). That is probably a good idea, with good intentions. However, it introduced conflicting requirements:

1. `Cache` should be extensible via the Module system and Cascading File System (as it always has been)
2. `Cache` should be available in `Kohana::cache()`, which is utilized before the Module system or CFS are initialized, in `Kohana::init()` (this is a new requirement via #24)

As I understand it, the decision has been made to drop requirement 1, because integrating `Cache` into the core system means that if I extend `Cache` in a custom module, or add a new driver in a module, it will *still not* be utilized during `init()` due to the fact that the Module system will not yet be initialized, and only the `SYSPATH` or `APPPATH` paths will be searched. Therefore, Koseven will `init()` with the core `Cache` class, save that path, and never use my extended version. Or worse, if caching is disabled, Koseven will use the core `Cache` class in `init()`, and my custom class in `shutdown_handler()`, which is the problem we're having now. Or throw an exception if it tries to load my custom driver.

It's the same original problem, except made more complex because instead of relying on the dumbed-down file-based caching built into `Kohana_Core`, several more classes and auto-loads will be involved.

I think the real solution to #24 is to change the utilization of `Kohana::cache` in other modules to `Cache::instance()`, and reserve `Kohana::cache` for usage only as a very basic `find_files()` cache.

If we want to enabled other caching mechanisms besides file-based caching for the `find_files()` cache, perhaps allowing a specific `cache` callable to be provided during `init()` would work? That makes it the job of the developer who wants this functionality to implement and they'll have the power to decide how they want to use Caching before Koseven is initialized.

Again, I don't want to undermine any work or thought that has already gone into it, I am just looking for some clarification on the path that has been chosen and how it addresses the above.

Thanks!

**Comment by augmenter** [comment]
Cache does not need to be integrated into the Core system. However it is quite questionable if we should cache the file structure into a centralized cache like Redis, all web nodes might not have the same directory layout.

I have fixed this issue locally by just modifying order of commands in the bootstrap:

- Kohana::modules($modules); //so Cache is loaded
- Cache::$default = 'opcache'; //or something else so we do not rely on FILE
- KO7::$config = new Config; //configure new instance of $config before Kohana::init call
- Kohana::$config->attach(new Config_File($config_directory)); //attach the config reader
- Kohana::init(...); // call Kohana init at the end so the Cache module will be already present

Looks simple but it took me several days of debugging to fix the find_file() caching, I had a lot of code fixes ready, but in the end this is the easiest fix.
